### PR TITLE
jax.scipy.special.owens_t — native Owen's T function

### LINKED ISSUE #37463: jax.scipy.special.owens_t — native Owen's T function
This is a follow-up to #12888, which requested Owen's T for computing the Skew Normal CDF. At the time, the suggested workaround was `tfp.math.owens_t` via `tensorflow_probability.substrates.jax`. Unfortunately, TFP-on-JAX is no longer actively maintained and is incompatible with recent JAX versions, so that path is no longer viable.

## Request

Add `jax.scipy.special.owens_t(h, a)` as a native special function, matching the `scipy.special.owens_t` interface.

## Motivation

Owen's T is required for an analytic Skew Normal CDF:

```
CDF(x; ξ, ω, α) = Φ((x − ξ) / ω) − 2 T((x − ξ) / ω, α)
```

This is needed for probabilistic modeling with Skew Normal likelihoods in JAX — without a native, JIT-compatible `owens_t`, there is no clean path to an analytic Skew Normal CDF.

## Implementation notes

TFP's existing implementation (`tensorflow_probability/python/math/special.py`) is a useful reference and translates fairly directly to JAX:

- It implements the full **Patefield & Tandy (2000)** algorithm (same as SciPy)
- The `(h, a)` plane is partitioned into **18 regions**, dispatched via `tf.where` chains — already branchless and vectorized, so it maps cleanly to `jnp.where`
- Six series expansion methods are used, each tuned for numerical stability in its region
- Custom VJP/JVP are registered because the region dispatch is not cleanly differentiable through autograd; the analytical partials are closed-form:
  - `∂T/∂h = −exp(−h²/2) · erf(ah/√2) / (2√(2π))`
  - `∂T/∂a = exp(−½(a²+1)h²) / (2π(a²+1))`

The main porting work is replacing `tf.math.*` / `tf.where` with `jnp.*` / `jnp.where`, and replacing TFP's custom gradient machinery with `jax.custom_vjp` / `jax.custom_jvp`. I would be happy to contribute an implementation.

## References
- Original request: #12888
- TFP implementation (reference): [`tensorflow_probability/python/math/special.py`](https://github.com/tensorflow/probability/blob/f05a4bbbdd34d2edb199b8ce8d0b755dacbad3b7/tensorflow_probability/python/math/special.py)
- SciPy: `scipy.special.owens_t`
- Patefield & Tandy (2000): http://www.jstatsoft.org/v05/i05/paper

- [x] Check for duplicate requests.

### PR DESCRIPTION
**Pull request by TheSkyentist**
## Summary

Closes #37463.

Adds `jax.scipy.special.owens_t(h, a)` as a native JAX special function, matching the `scipy.special.owens_t` interface.

This was motivated by the breakdown of the `tfp.math.owens_t` workaround — TFP-on-JAX is no longer actively maintained and is incompatible with recent JAX versions, leaving no clean path to an analytic Skew Normal CDF.

## Implementation

The implementation ports the Patefield & Tandy (2000) algorithm from TensorFlow Probability's `tensorflow_probability/python/math/special.py`, replacing TF ops with JAX equivalents:

- The `(h, a)` plane is partitioned into **18 regions** dispatched via `jnp.where` chains
- **6 numerical methods** are used (series expansions, Chebyshev approximation, 13-point Gaussian quadrature), selected by region
- The variable-length series loops (T1, T2, T4) are implemented as **Python `for` loops unrolled at trace time**, with data-dependent truncation via `jnp.where`. Max iterations are small (≤31), so unrolling is practical and avoids `lax.while_loop`
- **`jax.custom_jvp`** is registered with closed-form analytical partials, bypassing the non-differentiable `jnp.where` dispatch:
  - `∂T/∂h = −exp(−h²/2) · erf(ah/√2) / (2√(2π))`
  - `∂T/∂a = exp(−½(a²+1)h²) / (2π(a²+1))`

**Chebyshev coefficients (method T3):** The 31-coefficient array originates from the Boost Math library (`owens_t.hpp`, Benjamin Sobotta, 2012, Boost Software License 1.0) — specifically its 64-bit (long double) precision variant. Boost also provides a 22-coefficient variant targeting 53-bit (float64/double) precision, which would be sufficient for JAX's supported dtypes. We kept the 31-coefficient version to match TFP's behavior, but welcome feedback on whether to trim to 22.

## Tests

All 10 parametrised tests in `lax_scipy_special_functions_test.py` pass, exercising multiple shapes and dtypes across forward pass, JIT, and first-order gradients. Pre-commit hooks (ruff, Pyrefly) also pass.

Spot-checked accuracy:
- float32: max error ~3.6e-7 over 500 random samples in `[-3, 3]²`
- float64: max error ~6.9e-7 over 500 random samples in `[-4, 4]²`

## Notes / feedback requested

This PR is intentionally minimal — implementation and tests only. Before adding documentation we'd welcome feedback on:
- Whether the unrolled-loop approach is preferred over `lax.while_loop`
- The 22 vs 31 Chebyshev coefficient question above
- Any region boundary or edge-case concerns

*Code was written with the assistance of Claude Code; all outputs were author-verified.*

### CODE REVIEW

File: jax/_src/scipy/special.py [hunk 1]
```diff
@@ -3338,6 +3338,321 @@ def hyp2f1(a: ArrayLike, b: ArrayLike, c: ArrayLike, x: ArrayLike) -> Array:
...
 
+# Chebyshev series coefficients for Owen's T method T3.
+# Taken from the Boost Math library owens_t.hpp (Benjamin Sobotta, 2012),
+# Boost Software License 1.0, which provides two variants:
+#   22 coefficients for 53-bit (float64/double) precision, and
+#   31 coefficients for 64-bit (long double) precision (used here).
+# The 31-coefficient version is overkill for float64 but harmless.
+# Reference: Patefield & Tandy (2000), https://www.jstatsoft.org/v05/i05/paper
+_OWENS_T_CHEB_COEFFS = np.array([
+    0.99999999999999999999999729978162447266851932041876728736094298092,
```
Comment (jakevdp) [inline jax/_src/scipy/special.py]: Python floats are 64-bits, which are accurate to a max of ~17 significant digits. These constants have ~65 significant digits, so most will be truncated.

File: jax/_src/scipy/special.py [hunk 2]
```diff
@@ -3338,6 +3338,321 @@ def hyp2f1(a: ArrayLike, b: ArrayLike, c: ArrayLike, x: ArrayLike) -> Array:
...
+    -0.0607952766325955730493900985022020434830339794955745989150270485,
+    0.02163376832998715280598364838403905142754886795307972945570602292,
+    -0.0059340569345518672987699581418120390055001422042884348392721826,
+    0.00117434148183329465104745761827392105533338601068118659634858706,
+    -1.4891556133503689340734532606898813301663424844055299815106940e-4,
+    9.07235432079435758771092950798881466945428151426884488484154734e-6,
+])
+
+_OWENS_T_QUAD_PTS = np.array([
+    0.35082039676451715489e-02, 0.31279042338030753740e-01,
```
Comment (jakevdp) [inline jax/_src/scipy/special.py]: Even these are too many significant digits for float64.

File: jax/_src/scipy/special.py [hunk 3]
```diff
@@ -3338,6 +3338,321 @@ def hyp2f1(a: ArrayLike, b: ArrayLike, c: ArrayLike, x: ArrayLike) -> Array:
...
+    0.18042093461223385584e-01, 0.17263829606398753364e-01,
+    0.16243219975989856730e-01, 0.14994592034116704829e-01,
+    0.13535474469662088392e-01, 0.11886351605820165233e-01,
+    0.10070377242777431897e-01, 0.81130545742299586629e-02,
+    0.60419009528470238773e-02, 0.38862217010742057883e-02,
+    0.16793031084546090448e-02,
+])
+
+
+def _owens_t_method1(h, a, m):
```
Comment (jakevdp) [inline jax/_src/scipy/special.py]: It would be helpful if each of these mentioned in a comment which region they're applicable to.

File: jax/_src/scipy/special.py [hunk 4]
```diff
@@ -3338,6 +3338,321 @@ def hyp2f1(a: ArrayLike, b: ArrayLike, c: ArrayLike, x: ArrayLike) -> Array:
...
+
+  modified_a = jnp.where(abs_a <= 1., abs_a, jnp.reciprocal(abs_a))
+  modified_h = jnp.where(abs_a <= 1., h, abs_a * h)
+
+  regions = _owens_t_regions(modified_h, modified_a)
+
+  # Method 1 for regions 0-7 with data-dependent order
+  ones = jnp.ones_like(modified_h)
+  order1 = ones
+  order1 = jnp.where(regions[0], 2. * ones, order1)
```
Comment (jakevdp) [inline jax/_src/scipy/special.py]: Rather than a string of `where` clauses, please use `jax.lax.select_n` so that all cases can be handled in a single pass.

**Review by jakevdp — COMMENTED** [review]
Thanks for submitting this!

I worry that an implementation like this is going to be too slow to be useful in practice. Due to the way XLA lowers `where` / `select` statements, every branch is going to be fully evaluated for every input. That combined with the fact that individual branches generate code via large `for` loops, even evaluating this for a single scalar input will likely lead to very long compile times and runtimes. Even if the output is numerically correct, I'd rather not add a function with such poor performance to the core JAX library, as eventually a user is going to raise a bug related to its poor performance and we'll have no recourse.

Below I have a couple specific comments on the code, but none of those changes will address this fundamental issue, so it may not be worth your time to update the PR.

**Comment by TheSkyentist** [comment]
Thank you for the review! That totally makes sense, if every branch executes regardless, the whole benefit of the Patefield & Tandy (2000) algorithm is reversed; dispatching across 6 methods just multiplies the constant cost. However I still believe Owen's T function would be a useful function to port to `jax.scipy`. 

I propose that rather than following the original approach of PT2000 we could instead simply choose one branch over the entire domain (Method 5: GLQ). This would be 13 fixed points, no Python-level loop, clean vectorization, and excellent accuracy over the entire domain. At most I would add one branch to handle the asymptotically large case in h (Method 2) though I imagine the difference is tiny, but testable. 

﻿If that sounds reasonable, I'm happy to rewrite!

File: jax/_src/scipy/special.py (hunk 1 above)
Comment (TheSkyentist) [reply jax/_src/scipy/special.py]: These are removed in the new GLQ only mode

File: jax/_src/scipy/special.py (hunk 2 above)
Comment (TheSkyentist) [reply jax/_src/scipy/special.py]: I removed the extra sig figs.

File: jax/_src/scipy/special.py (hunk 3 above)
Comment (TheSkyentist) [reply jax/_src/scipy/special.py]: These are removed in the new GLQ only mode

File: jax/_src/scipy/special.py (hunk 4 above)
Comment (TheSkyentist) [reply jax/_src/scipy/special.py]: These are removed in the new GLQ only mode

**Comment by TheSkyentist** [comment]
I ended up having some time and pushed the GLQ only mode.

File: jax/_src/scipy/special.py [hunk 5]
```diff
@@ -3338,6 +3338,321 @@ def hyp2f1(a: ArrayLike, b: ArrayLike, c: ArrayLike, x: ArrayLike) -> Array:
...
+    s = jnp.where(i <= m, s + new_di * new_ai / (2. * i - 1.), s)
+    ai, di, gi = new_ai, new_di, new_gi
+  return s
+
+
+def _owens_t_method2(h, a, m):
+  h_sq = jnp.square(h)
+  nega_sq = -jnp.square(a)
+  y = jnp.reciprocal(h_sq)
+  summand = -0.5 * lax.erf(a * h) / h
```
[the remark this answers is not in the document]
Comment (TheSkyentist) [reply jax/_src/scipy/special.py]: Owen's method 2 is no longer a part of the code.

File: jax/_src/scipy/special.py [hunk 6]
```diff
@@ -3338,6 +3338,321 @@ def hyp2f1(a: ArrayLike, b: ArrayLike, c: ArrayLike, x: ArrayLike) -> Array:
...
+    summand, term = new_summand, new_term
+  return s * jnp.exp(-0.5 * h_sq) / np.sqrt(2 * np.pi)
+
+
+def _owens_t_method3(h, a):
+  a_sq = jnp.square(a)
+  h_sq = jnp.square(h)
+  y = jnp.reciprocal(h_sq)
+  vi = a * jnp.exp(-0.5 * jnp.square(a * h)) / np.sqrt(2 * np.pi)
+  zi = 0.5 * lax.erf(a * h / np.sqrt(2.)) / h
```
[the remark this answers is not in the document]
Comment (TheSkyentist) [reply jax/_src/scipy/special.py]: Owen's methods 2 and 3 are no longer a part of the code.

File: jax/_src/scipy/special.py [hunk 7]
```diff
@@ -3338,6 +3338,120 @@ def hyp2f1(a: ArrayLike, b: ArrayLike, c: ArrayLike, x: ArrayLike) -> Array:
...
+    0.010070377242777432, 0.0081130545742299587,
+    0.0060419009528470239, 0.0038862217010742058,
+    0.0016793031084546090,
+])
+
+
+def _owens_t_quadrature(h, a):
+  r = jnp.square(a)[..., None] * _OWENS_T_QUAD_PTS
+  integrand = jnp.exp(-0.5 * jnp.square(h)[..., None] * (1. + r)) / (1. + r)
+  return a * jnp.sum(integrand * _OWENS_T_QUAD_WTS, axis=-1)
```
Comment (jakevdp) [inline jax/_src/scipy/special.py]: Convert this sum to a batched matmul?

File: jax/_src/scipy/special.py [hunk 8]
```diff
@@ -3338,6 +3338,120 @@ def hyp2f1(a: ArrayLike, b: ArrayLike, c: ArrayLike, x: ArrayLike) -> Array:
...
+    0.016243219975989857, 0.014994592034116705,
+    0.013535474469662088, 0.011886351605820165,
+    0.010070377242777432, 0.0081130545742299587,
+    0.0060419009528470239, 0.0038862217010742058,
+    0.0016793031084546090,
+])
+
+
+def _owens_t_quadrature(h, a):
+  r = jnp.square(a)[..., None] * _OWENS_T_QUAD_PTS
```
Comment (jakevdp) [inline jax/_src/scipy/special.py]: This will trigger the implicit rank promotion warning. You can fix this by calling `expand_dims` on `_OWENS_T_QUAD_PTS`

File: jax/_src/scipy/special.py [hunk 9]
```diff
@@ -3338,6 +3338,120 @@ def hyp2f1(a: ArrayLike, b: ArrayLike, c: ArrayLike, x: ArrayLike) -> Array:
...
+          (0.25 - 0.25 * lax.erf(h / np.sqrt(2.)) * lax.erf(abs_a * h / np.sqrt(2.))
+           - result),
+          0.25 * (normh + normah - normh * normah) - result),
+      result)
+
+  result = lax.sign(a) * result
+  return jnp.where(jnp.isnan(a) | jnp.isnan(h), jnp.full_like(result, jnp.nan), result)
+
+
+@custom_derivatives.custom_jvp
```
Comment (jakevdp) [inline jax/_src/scipy/special.py]: I'd put this on `owens_t_impl`

**Review by jakevdp — CHANGES_REQUESTED** [review]
Looks good - a couple small changes.

Also, please add the new function to the list here, so it will appear in the documentation: https://github.com/jax-ml/jax/blob/4182dd052b0b12028583d0c7b117d2118b7245ff/docs/jax.scipy.rst?plain=1#L160-L166

**Comment by TheSkyentist** [comment]
I have done my best to address the comments :-)

File: jax/_src/scipy/special.py [hunk 10]
```diff
@@ -3338,6 +3338,121 @@ def hyp2f1(a: ArrayLike, b: ArrayLike, c: ArrayLike, x: ArrayLike) -> Array:
...
+  result = lax.sign(a) * result
+  return jnp.where(jnp.isnan(a) | jnp.isnan(h), jnp.full_like(result, jnp.nan), result)
+
+
+def _owens_t_jvp(primals, tangents):
+  (h, a) = primals
+  (dh, da) = tangents
+  result = _owens_t_impl(h, a)
+  # ∂T/∂h = -exp(-h²/2) · erf(ah/√2) / (2√(2π))
+  dout_dh = (-lax.exp(-0.5 * lax.square(h)) * lax.erf(a * h / np.sqrt(2.))
```
Comment (jakevdp) [inline jax/_src/scipy/special.py]: `np.sqrt(2.)` will return a `np.float64`, which could cause inadvertent type promotion here (e.g. if X64 mode is enabled but the inputs are float32). This might e.g. lead to slow 64-bit computations when running on a GPU.

A good fix here would be to use `np.sqrt(..., dtype=h.dtype)` in lines 3413 and 3414.

**Review by jakevdp — CHANGES_REQUESTED** [review]
Looks good! One last comment (sorry I didn't notice this the first time around).

Once this is fixed, please squash the changes into a single commit so we can merge.

Thanks for sticking through all the rounds of review!

**Comment by TheSkyentist** [comment]
Thank you for all of the feedback!

File: jax/_src/scipy/special.py [hunk 11]
```diff
@@ -3405,6 +3338,121 @@ def hyp2f1(a: ArrayLike, b: ArrayLike, c: ArrayLike, x: ArrayLike) -> Array:
...
+  modified_a = jnp.where(abs_a <= 1., abs_a, jnp.reciprocal(abs_a))
+  modified_h = jnp.where(abs_a <= 1., h, abs_a * h)
+
+  result = _owens_t_quadrature(modified_h, modified_a)
+
+  # Exact values for h=0 and a=1
+  result = jnp.where(modified_h == 0., arctan(modified_a) / (2 * np.pi), result)
+  result = jnp.where(
+      modified_a == 1.,
+      0.125 * lax.erfc(-modified_h / np.sqrt(2.)) * lax.erfc(modified_h / np.sqrt(2.)),
```
Comment (jakevdp) [inline jax/_src/scipy/special.py]: Sorry, I just noticed that there are other instances of `np.sqrt(2.)` here in the implementation. 

Given how frequently this comes up, I'd suggest defining a constant of the correct dtype just once:
```python
root_2 = _lax_const(h, np.sqrt(2))
```
and then refer to this throughout the function.

**Review by jakevdp — CHANGES_REQUESTED** [review]

File: jax/_src/scipy/special.py (hunk 11 above)
Comment (jakevdp) [reply jax/_src/scipy/special.py]: Or since `h / np.sqrt(2)` occurs so often, you could define `h_normed = h / np.sqrt(2, dtype=h.dtype)` and use that.

Comment (TheSkyentist) [reply jax/_src/scipy/special.py]: Good call! I defined the constants and updated the code.

**Review by jakevdp — APPROVED** [review]

**Comment by jakevdp** [comment]
Thanks for the contribution!
